File: packages/model-selector-py/src/model_selector/query/parser.py
"""Query string parser. Ported 1:1 from the TS implementation.

Query syntax:
- Comma-separated conditions, AND logic: ``"fast, cheap, functions"``
- Boolean: ``local``, ``!local``, ``functions``
- Comparisons: ``cost <= 5``, ``speed >= 7``, ``context_window >= 32000``
- Equality: ``provider = openai``, ``provider != google``
- Custom weights: ``local:10, fast:5``

Aliases are expanded before tokenizing. Position-based weighting gives the first
condition the highest weight by default.
"""

from __future__ import annotations

import re
from typing import Mapping

from ..errors import QueryParseError
from ..models import AttributeValue
from .types import ParsedQuery, QueryCondition

# Mirrors the TS regexes exactly.
_WEIGHT_RE = re.compile(r"^(.+):(\d+)$")
_COMPARISON_RE = re.compile(
    r"^([a-z_][a-z0-9_]*)\s*(>=|<=|!=|>|<|=)\s*(.+)$", re.IGNORECASE
)
_BOOL_ATTR_RE = re.compile(r"^[a-z_][a-z0-9_]*$", re.IGNORECASE)


def _to_number(raw: str) -> AttributeValue | None:
    """Parse a numeric literal, preferring int for integral values.

    Returns ``None`` when the string is not numeric (mirrors JS ``isNaN``).
    """
    try:
        return int(raw)
    except ValueError:
        pass
    try:
        return float(raw)
    except ValueError:
        return None


def parse_value(raw_value: str) -> AttributeValue:
    """Parse a value string into the appropriate type."""
    if raw_value == "true":
        return True
    if raw_value == "false":
        return False

    num = _to_number(raw_value)
    if num is not None:
        return num

    if (raw_value.startswith('"') and raw_value.endswith('"')) or (
        raw_value.startswith("'") and raw_value.endswith("'")
    ):
        return raw_value[1:-1]

    return raw_value


def _parse_token(token: str, position: int, total_conditions: int) -> QueryCondition:
    trimmed = token.strip()

    # Custom weight suffix (e.g. "local:10").
    weight = total_conditions - position  # default: position-based weight
    condition_part = trimmed

    weight_match = _WEIGHT_RE.match(trimmed)
    if weight_match:
        condition_part = weight_match.group(1).strip()
        weight = int(weight_match.group(2))

    # Negation prefix.
    negated = False
    if condition_part.startswith("!"):
        negated = True
        condition_part = condition_part[1:].strip()

    # Comparison operators.
    comparison_match = _COMPARISON_RE.match(condition_part)
    if comparison_match:
        attribute = comparison_match.group(1)
        operator = comparison_match.group(2)
        raw_value = comparison_match.group(3).strip()
        return QueryCondition(
            attribute=attribute,
            operator=operator,
            value=parse_value(raw_value),
            negated=negated,
            weight=weight,
        )

    # Simple boolean attribute (e.g. "local" or "!local").
    if _BOOL_ATTR_RE.match(condition_part):
        return QueryCondition(
            attribute=condition_part,
            operator="=",
            value=True,
            negated=negated,
            weight=weight,
        )

    raise QueryParseError(f'Invalid query condition: "{trimmed}"')


def expand_aliases(query: str, aliases: Mapping[str, str]) -> str:
    """Expand aliases in a query string before tokenizing."""
    tokens = [t.strip() for t in query.split(",")]
    expanded_tokens: list[str] = []

    for token in tokens:
        weight_match = _WEIGHT_RE.match(token)
        base_token = weight_match.group(1).strip() if weight_match else token
        weight_suffix = f":{weight_match.group(2)}" if weight_match else ""

        negated = base_token.startswith("!")
        lookup_token = base_token[1:] if negated else base_token

        alias_value = aliases.get(lookup_token)
        if alias_value:
            if negated:
                # For simple boolean aliases, negate; for comparison aliases this
                # may not be meaningful, but we still wrap so it parses.
                alias_value = f"!({alias_value})"
            expanded_tokens.append(alias_value + weight_suffix)
        else:
            expanded_tokens.append(token)

    return ", ".join(expanded_tokens)


def parse_query(query: str, aliases: Mapping[str, str] | None = None) -> ParsedQuery:
    """Parse a query string into a structured :class:`ParsedQuery`."""
    expanded_query = expand_aliases(query, aliases or {})

    tokens = [t.strip() for t in expanded_query.split(",")]
    tokens = [t for t in tokens if t]

    if not tokens:
        raise QueryParseError("Empty query")

    conditions = tuple(
        _parse_token(token, index, len(tokens)) for index, token in enumerate(tokens)
    )
    return ParsedQuery(conditions=conditions)

File: packages/model-selector-ts/src/query/parser.ts
import {
  AttributeValue,
  ComparisonOperator,
  ParsedQuery,
  QueryCondition,
  Aliases,
} from '../types.js';

/**
 * Parse a single query token into a condition.
 *
 * Supported formats:
 * - `attribute` -> attribute = true
 * - `!attribute` -> attribute = false
 * - `attribute = value` -> exact match
 * - `attribute >= 5` -> numeric comparison
 * - `attribute:weight` -> with custom weight (e.g., "local:10")
 */
function parseToken(token: string, position: number, totalConditions: number): QueryCondition {
  const trimmed = token.trim();

  // Check for custom weight suffix (e.g., "local:10")
  let weight = totalConditions - position; // Default: position-based weight
  let conditionPart = trimmed;

  const weightMatch = trimmed.match(/^(.+):(\d+)$/);
  if (weightMatch) {
    conditionPart = weightMatch[1]!.trim();
    weight = parseInt(weightMatch[2]!, 10);
  }

  // Check for negation prefix
  let negated = false;
  if (conditionPart.startsWith('!')) {
    negated = true;
    conditionPart = conditionPart.slice(1).trim();
  }

  // Try to parse comparison operators
  const comparisonMatch = conditionPart.match(
    /^([a-z_][a-z0-9_]*)\s*(>=|<=|!=|>|<|=)\s*(.+)$/i
  );

  if (comparisonMatch) {
    const attribute = comparisonMatch[1]!;
    const operator = comparisonMatch[2] as ComparisonOperator;
    const rawValue = comparisonMatch[3]!.trim();
    const value = parseValue(rawValue);

    return {
      attribute,
      operator,
      value,
      negated,
      weight,
    };
  }

  // Simple boolean attribute (e.g., "local" or "!local")
  if (/^[a-z_][a-z0-9_]*$/i.test(conditionPart)) {
    return {
      attribute: conditionPart,
      operator: '=',
      value: true,
      negated,
      weight,
    };
  }

  throw new Error(`Invalid query condition: "${trimmed}"`);
}

/**
 * Parse a value string into the appropriate type.
 */
function parseValue(rawValue: string): AttributeValue {
  // Boolean
  if (rawValue === 'true') return true;
  if (rawValue === 'false') return false;

  // Number
  const num = Number(rawValue);
  if (!isNaN(num)) return num;

  // String (remove quotes if present)
  if ((rawValue.startsWith('"') && rawValue.endsWith('"')) ||
      (rawValue.startsWith("'") && rawValue.endsWith("'"))) {
    return rawValue.slice(1, -1);
  }

  return rawValue;
}

/**
 * Expand aliases in a query string.
 */
function expandAliases(query: string, aliases: Aliases): string {
  let expanded = query;

  // Split by comma, expand each token, rejoin
  const tokens = expanded.split(',').map((t) => t.trim());
  const expandedTokens = tokens.map((token) => {
    // Check if this token (without weight suffix) is an alias
    const weightMatch = token.match(/^(.+):(\d+)$/);
    const baseToken = weightMatch ? weightMatch[1]!.trim() : token;
    const weightSuffix = weightMatch ? `:${weightMatch[2]}` : '';

    // Check for negation
    const negated = baseToken.startsWith('!');
    const lookupToken = negated ? baseToken.slice(1) : baseToken;

    if (aliases[lookupToken]) {
      // Expand the alias
      let aliasValue = aliases[lookupToken]!;
      // If negated, we need to handle this specially
      if (negated) {
        // For simple boolean aliases, negate them
        // For comparison aliases, this might not make sense - just prepend !
        aliasValue = `!(${aliasValue})`;
      }
      return aliasValue + weightSuffix;
    }

    return token;
  });

  return expandedTokens.join(', ');
}

/**
 * Parse a query string into a structured ParsedQuery.
 *
 * Query syntax:
 * - Comma-separated conditions (AND logic)
 * - Boolean: `local`, `!local`, `functions`
 * - Comparisons: `cost <= 5`, `speed >= 7`, `context_window >= 32000`
 * - Equality: `provider = openai`
 * - Custom weights: `local:10, fast:5`
 *
 * Position-based weighting: first condition gets highest weight by default.
 */
export function parseQuery(query: string, aliases: Aliases = {}): ParsedQuery {
  // Expand aliases first
  const expandedQuery = expandAliases(query, aliases);

  // Split by comma
  const tokens = expandedQuery.split(',').map((t) => t.trim()).filter((t) => t.length > 0);

  if (tokens.length === 0) {
    throw new Error('Empty query');
  }

  const conditions = tokens.map((token, index) =>
    parseToken(token, index, tokens.length)
  );

  return { conditions };
}
